File: examples/async.rs
#[calcite::deno_op]
async fn sleep_for(secs: i32) -> Result<String, &'static str> {
    if secs < 0 {
        Err("Number of seconds cannot be negative")
    } else {
        let (tx, rx) = futures::channel::oneshot::channel();
        std::thread::spawn(move || {
            std::thread::sleep(std::time::Duration::from_secs(secs as u64));
            tx.send(()).unwrap();
        });
        rx.await.unwrap();
        Ok(format!("Slept for {} seconds", secs))
    }
}

calcite::export!(sleep_for);

File: examples/async.ts
import {
  loadPlugin,
  importAsyncFromPlugin,
} from "https://deno.land/x/calcite@2.1/calcite.ts";
await loadPlugin("async", "file://target/debug/examples/");

const sleep_for = importAsyncFromPlugin("sleep_for") as (
  secs: number,
) => Promise<string>;

sleep_for(5).then((response) => console.log(response));
sleep_for(3).then((response) => console.log(response));
sleep_for(-1).catch((err) => console.log(`Got error ${err}`));
